### PR TITLE
ATS9440

### PR DESCRIPTION
**Pull request by mizkulg**
-Updated spacing. 

@jenshnielsen @FarBo

### CODE REVIEW

**Comment by jenshnielsen** [comment]
@mizkulg There is no need to open a new pull request every time you make changes. You can push a new commit to the same branch.

**Comment by FarBo** [comment]
Thanks for the PR @mizkulg 
I am in the process of reviewing it.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+                                               'SLOW_EXTERNAL_CLOCK': 4,
+                                               'EXTERNAL_CLOCK_10MHz_REF': 7})
+        self.add_parameter(name='external_sample_rate',
+                           get_cmd=None,
+                           parameter_class=TraceParameter,
+                           label='External Sample Rate',
+                           unit='S/s',
+                           vals=validators.MultiType(validators.Ints(1000000, 125000000),
+                                                     validators.Enum('UNDEFINED')),
+                           initial_value='UNDEFINED')
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:37]: The card supports slow external clock for the range below 1 MHz.
https://www.alazartech.com/Support/Download%20Files/ATS9440%20User%20Manual_V1_0.pdf (page 38).
I think this future should be implemented or specified as TODO to be done in the future.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+                           val_mapping={'CLOCK_EDGE_RISING': 0,
+                                        'CLOCK_EDGE_FALLING': 1})
+        self.add_parameter(name='decimation',
+                           get_cmd=None,
+                           parameter_class=TraceParameter,
+                           label='Decimation',
+                           unit=None,
+                           initial_value=1,
+                           vals=validators.Ints(1, 100000))
+        for i in ['1', '2']:
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:78]: This is a loop over the channels so it should be `['1', '2', '3', '4']`

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+                           unit=None,
+                           initial_value='TRIG_ENGINE_OP_J',
+                           val_mapping={'TRIG_ENGINE_OP_J': 0,
+                                        'TRIG_ENGINE_OP_K': 1,
+                                        'TRIG_ENGINE_OP_J_OR_K': 2,
+                                        'TRIG_ENGINE_OP_J_AND_K': 3,
+                                        'TRIG_ENGINE_OP_J_XOR_K': 4,
+                                        'TRIG_ENGINE_OP_J_AND_NOT_K': 5,
+                                        'TRIG_ENGINE_OP_NOT_J_AND_K': 6})
+        for i in ['1', '2']:
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:122]: This is a loop over the triggers so it should stay as is

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+                           unit='S/s',
+                           vals=validators.MultiType(validators.Ints(1000000, 125000000),
+                                                     validators.Enum('UNDEFINED')),
+                           initial_value='UNDEFINED')
+        self.add_parameter(name='sample_rate',
+                           get_cmd=None,
+                           parameter_class=TraceParameter,
+                           label='Internal Sample Rate',
+                           unit='S/s',
+                           initial_value=100000000,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:43]: Is there any specific reason for having this initial value?

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+                                    2_000_000: 24,
+                                    5_000_000: 26,
+                                   10_000_000: 28,
+                                   20_000_000: 30,
+                                   50_000_000: 34,
+                                  100_000_000: 36,
+                                  125_000_000: 38,
+                             'EXTERNAL_CLOCK': 64,
+                                  'UNDEFINED': 'UNDEFINED'})
+        self.add_parameter(name='clock_edge',
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:63]: This parameter can be set (it can have set_cmd=None)

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+                                        'CLOCK_EDGE_FALLING': 1})
+        self.add_parameter(name='decimation',
+                           get_cmd=None,
+                           parameter_class=TraceParameter,
+                           label='Decimation',
+                           unit=None,
+                           initial_value=1,
+                           vals=validators.Ints(1, 100000))
+        for i in ['1', '2']:
+            self.add_parameter(name='coupling' + i,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:79]: This parameter is also settable between 'AC' and 'DC'
https://www.alazartech.com/Support/Download%20Files/ATS9440%20User%20Manual_V1_0.pdf (page 51)

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+                               get_cmd=None,
+                               parameter_class=TraceParameter,
+                               label='Coupling channel ' + i,
+                               unit=None,
+                               initial_value='DC',
+                               val_mapping={'AC': 1, 'DC': 2})
+            self.add_parameter(name='channel_range' + i,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:86]: This parameter is also settable between 6 different values. 
https://www.alazartech.com/Support/Download%20Files/ATS9440%20User%20Manual_V1_0.pdf (page 51)

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+                               get_cmd=None,
+                               parameter_class=TraceParameter,
+                               label='Range channel ' + i,
+                               unit='V',
+                               initial_value=0.4,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:91]: initial_value=0.1 (it maybe is a good idea to set the initial value to the minimum)

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 2 above)
Comment (jenshnielsen) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:78]: @mizkulg

Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:78]: Done and compiles.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 4 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:43]: No, I just set it somewhere between the min and max.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 5 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:63]: What do you mean be set? I'm not seeing where the problem is here.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 6 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:79]: So it should be 'val_mapping={'DC' 1}`?

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 5 above)
Comment (jenshnielsen) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:63]: Note that for `clock_edge` you do not do `set_cmd=None`, this is done for most other paramters. If you do not do that then `set_cmd=False` which means that you cannot set `clock_edge` only get it.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 6 above)
Comment (jenshnielsen) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:79]: No as above it should also have `set_cmd=None`

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 8 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:91]: changed

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 7 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:86]: Not seeing these six values.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 5 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:63]: So just add `set_cmd=None` to `clock_edge`?

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 6 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:79]: Ok added `set_cmd=None`

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 7 above)
Comment (FarBo) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:86]: On the page 51, there is "Full Scale input ranges", which is from +-100 mV up to +-4 V. For ATS9360, this was only +-0.4 V. So, there should be set_cmd=None for this parameter. val_mapping should be also modified, but I am not sure how!

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+                               parameter_class=TraceParameter,
+                               label='Trigger Source ' + i,
+                               unit=None,
+                               initial_value='EXTERNAL',
+                               val_mapping={'CHANNEL_A': 0,
+                                            'CHANNEL_B': 1,
+                                            'EXTERNAL': 2,
+                                            'DISABLE': 3,
+                                            'CHANNEL_C': 4,
+                                            'CHANNEL_D': 5})
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:142]: Maybe this is a bit cleaner:
val_mapping={'CHANNEL_A': 0,
                                            'CHANNEL_B': 1,
                                            'CHANNEL_C': 2,
                                            'CHANNEL_D': 3,
                                            'EXTERNAL': 4,
                                            'DISABLE': 5,})

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 7 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:86]: I'm also not sure how the six values should be defined. Perhaps @jenshnielsen could help here. There has to be a reason why `0.4: 7`is for ATS9360. We can then add that structure to the six values for ATS9440.

Comment (jenshnielsen) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:86]: The supported ranges are listed here https://www.alazartech.com/Support/Download%20Files/ATS-SDK-Guide.pdf in chapter 7.1 (page 185) and the values needed to set them on page 150.-153

Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:86]: Thanks  @jenshnielsen. Page helped. Updated the code.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 10]
```diff
@@ -136,10 +146,10 @@ def __init__(self, name, **kwargs):
...
                                             'CHANNEL_B': 1,
-                                            'EXTERNAL': 2,
-                                            'DISABLE': 3,
-                                            'CHANNEL_C': 4,
-                                            'CHANNEL_D': 5})
+                                            'CHANNEL_C': 2,
+                                            'CHANNEL_D': 3,
+                                            'EXTERNAL': 4,
+                                            'DISABLE': 5})
             self.add_parameter(name='trigger_slope' + i,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:153]: The previous order is correct. i commented wrong and I erased it.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 11]
```diff
@@ -119,7 +129,7 @@ def __init__(self, name, **kwargs):
                                         'TRIG_ENGINE_OP_J_XOR_K': 4,
                                         'TRIG_ENGINE_OP_J_AND_NOT_K': 5,
                                         'TRIG_ENGINE_OP_NOT_J_AND_K': 6})
-        for i in ['1', '2']:
+        for i in ['1', '2','3','4']:
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:132]: Trigger is only 2 iterations, exactly like ATS9360. No need to change it to 4.  Only the loop for channels needs to be changed to 4 iterations as we have 4 channels (you have correctly changed it above). Please revert this change back.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 12]
```diff
@@ -11,7 +11,7 @@ class AlazarTech_ATS9440(AlazarTech_ATS):
     samples_divisor = 256 #32
 
     def __init__(self, name, **kwargs):
-        dll_path = 'C:\\WINDOWS\\System32\\ATSApi.dll'
+        dll_path = '/usr/lib64/libATSApi.so'
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:14]: I think you were trying the driver on Linux. As we want to stay on Windows, you should keep the windows API address.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 13]
```diff
@@ -75,21 +76,30 @@ def __init__(self, name, **kwargs):
...
                                initial_value='DC',
                                val_mapping={'AC': 1, 'DC': 2})
             self.add_parameter(name='channel_range' + i,
                                get_cmd=None,
+                               set_cmd=None,
                                parameter_class=TraceParameter,
                                label='Range channel ' + i,
                                unit='V',
-                               initial_value=0.4,
-                               val_mapping={0.4: 7})
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py]: This can be erased here: val_mapping={0.4: 7})

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 11 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:132]: Yea my mistake. Wasn't supposed to be changed. Was just curious what the errors would be.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 12 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:14]: Yea that was just for my personal testing. Changed it back to Windows.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+    def __init__(self, name, **kwargs):
+        dll_path = 'C:\\WINDOWS\\System32\\ATSApi.dll'
+        super().__init__(name, dll_path=dll_path, **kwargs)
+
+        # add parameters
+
+        # ----- Parameters for the configuration of the board -----
+        self.add_parameter(name='clock_source',
+                           parameter_class=TraceParameter,
+                           get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:22]: set_cmd=None can be added to this parameter.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+                           label='Clock Source',
+                           unit=None,
+                           initial_value='INTERNAL_CLOCK',
+                           val_mapping={'INTERNAL_CLOCK': 1,
+                                               'FAST_EXTERNAL_CLOCK': 2,
+                                               'SLOW_EXTERNAL_CLOCK': 4,
+                                               'EXTERNAL_CLOCK_10MHz_REF': 7})
+        self.add_parameter(name='external_sample_rate',
+                           get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:31]: set_cmd=None can be added to this parameter.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+                           parameter_class=TraceParameter,
+                           label='External Sample Rate',
+                           unit='S/s',
+                           vals=validators.MultiType(validators.Ints(1000000, 125000000),
+                                                     validators.Enum('UNDEFINED')),
+                           initial_value='UNDEFINED')
+        self.add_parameter(name='sample_rate',
+                           get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:39]: set_cmd=None can be added to this parameter.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+                               parameter_class=TraceParameter,
+                               label='Range channel ' + i,
+                               unit='V',
+                               initial_value=0.1,
+                               val_mapping={0.1: 5,
+                                            0.2: 6,
+                                            0.4: 7,
+                                            1: 10,
+                                            2: 11,
+                                            4: 12,})
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:100]: You can erase , at the end

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 18]
```diff
@@ -0,0 +1,328 @@
...
+            self.add_parameter(name='impedance' + i,
+                               get_cmd=None,
+                               parameter_class=TraceParameter,
+                               label='Impedance channel ' + i,
+                               unit='Ohm',
+                               initial_value=50,
+                               val_mapping={50: 2})
+
+            self.add_parameter(name='bwlimit' + i,
+                               get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:110]: This can also have set_cmd=None

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+                               parameter_class=TraceParameter,
+                               label='Bandwidth limit channel ' + i,
+                               unit=None,
+                               initial_value='DISABLED',
+                               val_mapping={'DISABLED': 0,
+                                            'ENABLED': 1})
+        self.add_parameter(name='trigger_operation',
+                           get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:118]: This can also have set_cmd=None

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 20]
```diff
@@ -0,0 +1,328 @@
...
+            self.add_parameter(name='trigger_engine' + i,
+                               get_cmd=None,
+                               parameter_class=TraceParameter,
+                               label='Trigger Engine ' + i,
+                               unit=None,
+                               initial_value='TRIG_ENGINE_' + ('J' if i == '1' else 'K'),
+                               val_mapping={'TRIG_ENGINE_J': 0,
+                                            'TRIG_ENGINE_K': 1})
+            self.add_parameter(name='trigger_source' + i,
+                               get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:140]: set_cmd=None for this one as well.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 21]
```diff
@@ -0,0 +1,328 @@
...
+                               unit=None,
+                               initial_value='EXTERNAL',
+                               val_mapping={'CHANNEL_A': 0,
+                                            'CHANNEL_B': 1,
+                                            'EXTERNAL': 2,
+                                            'DISABLE': 3,
+                                            'CHANNEL_C': 4,
+                                            'CHANNEL_D': 5})
+            self.add_parameter(name='trigger_slope' + i,
+                               get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:152]: set_cmd=None

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 22]
```diff
@@ -0,0 +1,328 @@
...
+                               parameter_class=TraceParameter,
+                               label='Trigger Slope ' + i,
+                               unit=None,
+                               initial_value='TRIG_SLOPE_POSITIVE',
+                               val_mapping={'TRIG_SLOPE_POSITIVE': 1,
+                                            'TRIG_SLOPE_NEGATIVE': 2})
+            self.add_parameter(name='trigger_level' + i,
+                               get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:160]: set_cmd=None

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 14 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:22]: Done

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 23]
```diff
@@ -0,0 +1,328 @@
...
+                               vals=validators.Ints(0, 255))
+        self.add_parameter(name='external_trigger_coupling',
+                           get_cmd=None,
+                           parameter_class=TraceParameter,
+                           label='External Trigger Coupling',
+                           unit=None,
+                           initial_value='DC',
+                           val_mapping={'AC': 1,'DC': 2})
+        self.add_parameter(name='external_trigger_range',
+                           get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:174]: set_cmd=None

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 24]
```diff
@@ -0,0 +1,328 @@
...
+                           parameter_class=TraceParameter,
+                           label='External Trigger Range',
+                           unit=None,
+                           initial_value='ETR_5V',
+                           val_mapping={'ETR_5V': 0,'ETR_TTL': 2})
+        self.add_parameter(name='trigger_delay',
+                           get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:181]: set_cmd=None

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 25]
```diff
@@ -0,0 +1,328 @@
...
+                           parameter_class=TraceParameter,
+                           label='Trigger Delay',
+                           unit='Sample clock cycles',
+                           initial_value=0,
+                           vals=validators.Multiples(divisor=8, min_value=0))
+        self.add_parameter(name='timeout_ticks',
+                           get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:188]: set_cmd=None

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 26]
```diff
@@ -0,0 +1,340 @@
...
+                           parameter_class=TraceParameter,
+                           label='AUX I/O Mode',
+                           unit=None,
+                           initial_value='AUX_IN_AUXILIARY',
+                           val_mapping={'AUX_OUT_TRIGGER': 0,
+                                        'AUX_IN_TRIGGER_ENABLE': 1,
+                                        'AUX_IN_AUXILIARY': 13})
+        self.add_parameter(name='aux_io_param',
+                           get_cmd=None,
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:216]: set_cmd=None

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 27]
```diff
@@ -0,0 +1,340 @@
...
+                           initial_value=0,
+                           vals=validators.Multiples(divisor=8, min_value=0))
+        self.add_parameter(name='timeout_ticks',
+                           get_cmd=None,
+                           set_cmd=None,
+                           parameter_class=TraceParameter,
+                           label='Timeout Ticks',
+                           unit='10 us',
+                           initial_value=0,
+                           vals=validators.Ints(min_value=0))
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:205]: Such a comment should be added:
# The card has two AUX I/O ports, which AUX 2 is only controlled by the driver. AUX 1 is controlled by the firmware. So,
# the user should use AUX 2 to control it via aux_io_mode and aux-io_param.

Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:205]: Such a comment should be added here:
The card has two AUX I/O ports, which only AUX 2 is controlled by the software (AUX 1 is controlled by the firmware). 
The user should use AUX 2 for controlling the AUX via aux_io_mode and aux_io_param.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 28]
```diff
@@ -0,0 +1,340 @@
...
+        self.add_parameter(name='aux_io_mode',
+                           get_cmd=None,
+                           parameter_class=TraceParameter,
+                           label='AUX I/O Mode',
+                           unit=None,
+                           initial_value='AUX_IN_AUXILIARY',
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:212]: As far as I can see in the manual, the default value is TRIG OUT. So, 
initial_value='AUX_OUT_TRIGGER'
https://www.alazartech.com/Support/Download%20Files/ATS9440%20User%20Manual_V1_0.pdf (Page 31)

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 29]
```diff
@@ -0,0 +1,341 @@
...
+                                        'TRIG_SLOPE_NEGATIVE': 2})
+
+        #The above parameters are important for preparing the card.
+        self.add_parameter(name='mode',
+                           label='Acquisition mode',
+                           unit=None,
+                           initial_value='NPT',
+                           get_cmd=None,
+                           set_cmd=None,
+                           val_mapping={'NPT': 0x200, 'TS': 0x400})
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:233]: We have 4 Acquisition modes for the card:
val_mapping={'TM':0x000, 'CM':0x100, 'NPT': 0x200, 'TS': 0x400}

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py [hunk 30]
```diff
@@ -0,0 +1,341 @@
...
+from .utils import TraceParameter
+from qcodes.utils import validators
+
+
+class AlazarTech_ATS9440(AlazarTech_ATS):
+    """
+    This class is the driver for the ATS9440 board
+    it inherits from the ATS base class
+    """
+    samples_divisor = 256 #32
```
Comment (FarBo) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:11]: This should be 32

Comment (jenshnielsen) [inline qcodes/instrument_drivers/AlazarTech/ATS9440.py:11]: Please add a line line `channels = 4` here so that you tell the driver that this card has 4 channels

Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:11]: I had 32 as seen but I kept getting an error that is has to be greater than 256. I'll change it back for 32 for now and will check again if I get the same error after all the changes we made.

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 27 above)
Comment (mizkulg) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:205]: So just a comment under the aux_io_mode parameter?

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 29 above)
Comment (jenshnielsen) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:233]: We don't expose all modes on the other cards so I think this is fine to leave as is

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 26 above)
Comment (jenshnielsen) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:217]: Not needed

File: qcodes/instrument_drivers/AlazarTech/ATS9440.py (hunk 27 above)
Comment (jenshnielsen) [reply qcodes/instrument_drivers/AlazarTech/ATS9440.py:205]: Just here before aux_io_mode would be fine

**Review by jenshnielsen — APPROVED** [review]
Thanks, @FarBo  I think this is ready to land. Please have a final look

**Review by FarBo — APPROVED** [review]
